nav2_behavior_tree/test/plugins/action/test_clear_costmap_service.cpp

```diff
@@ -0,0 +1,130 @@
+#include <set>
+#include <string>
+
+#include "behaviortree_cpp_v3/bt_factory.h"
+
+#include "../../test_service.hpp"
+#include "nav2_behavior_tree/plugins/action/clear_costmap_service.hpp"
+
+class ClearEntireCostmapService : public TestService<nav2_msgs::srv::ClearEntireCostmap>
+{
+public:
+  ClearEntireCostmapService()
+  : TestService("clear_entire_costmap")
+  {}
+};
+
+class ClearEntireCostmapServiceTestFixture : public ::testing::Test
+{
+public:
+  static void SetUpTestCase()
+  {
+    node_ = std::make_shared<rclcpp::Node>("clear_entire_costmap_test_fixture");
+    factory_ = std::make_shared<BT::BehaviorTreeFactory>();
+
+    config_ = new BT::NodeConfiguration();
+
+    // Create the blackboard that will be shared by all of the nodes in the tree
+    config_->blackboard = BT::Blackboard::create();
+    // Put items on the blackboard
+    config_->blackboard->set<rclcpp::Node::SharedPtr>(
+      "node",
+      node_);
+    config_->blackboard->set<std::chrono::milliseconds>(
+      "server_timeout",
+      std::chrono::milliseconds(10));
+    config_->blackboard->set<bool>("path_updated", false);
+    config_->blackboard->set<bool>("initial_pose_received", false);
+    config_->blackboard->set<int>("number_recoveries", 0);
+
+    factory_->registerNodeType<nav2_behavior_tree::ClearEntireCostmapService>("ClearEntireCostmap");
+  }
+
+  static void TearDownTestCase()
+  {
+    delete config_;
+    config_ = nullptr;
+    node_.reset();
+    server_.reset();
+    factory_.reset();
+  }
+
+  void SetUp() override
+  {
+    config_->blackboard->set("number_recoveries", 0);
+  }
+
+  void TearDown() override
+  {
+    tree_.reset();
+  }
+
+  static std::shared_ptr<ClearEntireCostmapService> server_;
+
+protected:
+  static rclcpp::Node::SharedPtr node_;
+  static BT::NodeConfiguration * config_;
+  static std::shared_ptr<BT::BehaviorTreeFactory> factory_;
+  static std::shared_ptr<BT::Tree> tree_;
+};
+
+rclcpp::Node::SharedPtr ClearEntireCostmapServiceTestFixture::node_ = nullptr;
+std::shared_ptr<ClearEntireCostmapService> ClearEntireCostmapServiceTestFixture::server_ = nullptr;
+BT::NodeConfiguration * ClearEntireCostmapServiceTestFixture::config_ = nullptr;
+std::shared_ptr<BT::BehaviorTreeFactory> ClearEntireCostmapServiceTestFixture::factory_ = nullptr;
+std::shared_ptr<BT::Tree> ClearEntireCostmapServiceTestFixture::tree_ = nullptr;
+
+TEST_F(ClearEntireCostmapServiceTestFixture, test_tick)
+{
+  std::string xml_txt =
+    R"(
+      <root main_tree_to_execute = "MainTree" >
+        <BehaviorTree ID="MainTree">
+            <ClearEntireCostmap service_name="clear_entire_costmap"/>
+        </BehaviorTree>
+      </root>)";
+
+  tree_ = std::make_shared<BT::Tree>(factory_->createTreeFromText(xml_txt, config_->blackboard));
+  EXPECT_EQ(config_->blackboard->get<int>("number_recoveries"), 0);
+  EXPECT_EQ(tree_->rootNode()->executeTick(), BT::NodeStatus::SUCCESS);
+  EXPECT_EQ(config_->blackboard->get<int>("number_recoveries"), 1);
```

Comment:
I feel like I want more from this, but not sure what else you could do. By this variable incrementing, we know it was called and did something, but I suppose there's no closing the loop.

Reply:
Ah yes I get what you mean. But this was all I could actually test with the service nodes while being independent from everything else. Again, I definitely agree there should be a proper test to check if the costmap is being cleared correctly, but I'm not sure it belongs here.

Reply:
fair enough. We have a recoveries section of nav2_system_tests for the action based ones, but we never test the costmaps, which we should. I'll file a ticket